pkg/snapshot/restorer/restorer.go

```diff
@@ -317,25 +326,192 @@ func startEmbeddedEtcd(ro RestoreOptions) (*embed.Etcd, error) {
 	return e, nil
 }
 
-// applyDeltaSnapshot applies the events from time sorted list of delta snapshot to etcd sequentially
+// applyDeltaSnapshots fetches the events from time sorted list of delta snapshots parallelly and applies them to etcd pseudo-sequentially
 func (r *Restorer) applyDeltaSnapshots(client *clientv3.Client, snapList snapstore.SnapList) error {
+	startTime := time.Now()
+	r.logger.Infof("--- APPLY DELTA SNAPSHOTS")
+
+	for _, snap := range snapList {
+		r.logger.Infof("--- APPLY DELTA SNAPSHOTS - snap revision = %d, %d", snap.StartRevision, snap.LastRevision)
+	}
+
 	firstDeltaSnap := snapList[0]
 	if err := r.applyFirstDeltaSnapshot(client, *firstDeltaSnap); err != nil {
 		return err
 	}
 	if err := verifySnapshotRevision(client, snapList[0]); err != nil {
 		return err
 	}
-	for _, snap := range snapList[1:] {
-		if err := r.applyDeltaSnapshot(client, *snap); err != nil {
-			return err
+
+	r.logger.Infof("--- APPLY DELTA SNAPSHOTS - first delta snapshot applied")
+
+	errCh := make(chan error, 1)
+
+	remainingSnaps := snapList[1:]                // stores the snaps, excluding the first one
+	fetchedEventsInfo := make(chan eventsInfo, 1) // channel that passes fetched event data from snaps
+	restoreFinished := make(chan bool, 1)         // channel used to indicate that restoration is finished
+
+	r.logger.Infof("--- APPLY DELTA SNAPSHOTS - channels created")
+
+	// snapLocations is an array to record pointers to temp file locations to which the events objects are stored as json strings.
+	snapLocations := make([]string, len(remainingSnaps))
+
+	go r.fetchListener(snapLocations, fetchedEventsInfo, errCh)
+	r.logger.Infof("--- APPLY DELTA SNAPSHOTS - fetchListener goroutine started")
+	go r.startFetchers(remainingSnaps, snapLocations, fetchedEventsInfo, errCh)
+	r.logger.Infof("--- APPLY DELTA SNAPSHOTS - startFetchers goroutine started")
+	go r.applier(client, remainingSnaps, snapLocations, restoreFinished, errCh)
+	r.logger.Infof("--- APPLY DELTA SNAPSHOTS - applier goroutine started")
+
+	select {
+	case <-restoreFinished:
+		r.logger.Infof("--- APPLY DELTA SNAPSHOTS - select restoreFinished")
+		endTime := time.Now()
+		restorationTime := endTime.Sub(startTime)
+		r.logger.Warnf("RESTORATION TIME: %v", restorationTime)
+		return nil
+	case err := <-errCh:
+		r.logger.Infof("--- APPLY DELTA SNAPSHOTS - select errCh")
+		return err
+	}
+}
+
+// coordinator coordinates the actions of the fetcher routines and the applier routine. It ensures that events are applied in the right order, regardless of the order in which snaps are fetched
```

Comment:
The events are persisted in the files. I have one concern, though, with the implementation. Say I have 5 incr snaps and I fetch the 5 snaps in 5 threads. Now if say, 3 failed and others succeeded, applier will apply 1 and 2. Files are created for 4 and 5. 3 returned an err so files 4 and 5 are not cleaned up. 

Another problem I see is that, after the initializer starts after the err and does the data validation check, it will see that data directory is valid, though we have not applied 3, 4 and 5. That will be catastrophic. We need to remove contents from the data directory if restore fails.

Reply:
Thanks for the heads up about file cleanup! I completely agree!

I see that you've raised a separate PR for the data directory cleanup for failed restores. I hope there's nothing to do for that from my side, right? Just confirming.